jax_md/elasticity.py

```diff
@@ -0,0 +1,566 @@
+
+  if rank == 1:
+    def extract(i, j):
+      idx = mandel_index(i, j)
+      return M[idx] / weight[idx]
+    T = vmap(
+          vmap(extract, 
+            in_axes=(0,None)), 
+          in_axes=(None,0))(tensor_range, tensor_range)
+  else:
+    def extract(i, j, k, l):
+      idx0 = mandel_index(i, j)
+      idx1 = mandel_index(k, l)
+      return M[idx0, idx1] / (weight[idx0] * weight[idx1])
+    T = vmap(
+          vmap(
+            vmap(
+              vmap(extract, 
+                in_axes=(0,None,None,None)), 
+              in_axes=(None,0,None,None)), 
+            in_axes=(None,None,0,None)), 
+          in_axes=(None,None,None,0))(tensor_range, 
+                                      tensor_range, 
+                                      tensor_range, 
+                                      tensor_range)
+  return T
+
+
+
+@partial(jit,static_argnums=(1,))
+def _extract_elements(C, as_dict):
+  if C.shape[0] == 2:
+    indices = ( [0, 1, 0, 0, 0, 0],
+                [0, 1, 1, 0, 0, 1],
+                [0, 1, 0, 1, 0, 1],
+                [0, 1, 1, 1, 1, 1])
+    clist = C[ indices ] 
+    if as_dict:
+      names = ['cxxxx','cyyyy','cxyxy','cxxyy','cxxxy','cyyxy']
+      return dict(zip(names, clist))
+    else:
+      return clist
+
+  elif C.shape[0] == 3:
+    indices = ( [0, 1, 2, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0],
+                [0, 1, 2, 2, 2, 1, 1, 0, 0, 0, 0, 0, 1, 2, 1, 2, 2, 1, 2, 1, 1],
+                [0, 1, 2, 1, 0, 0, 2, 2, 1, 1, 0, 0, 1, 1, 1, 2, 2, 2, 1, 1, 0],
+                [0, 1, 2, 2, 2, 1, 2, 2, 1, 2, 2, 1, 2, 1, 1, 2, 2, 2, 2, 2, 2])
+    clist = C[ indices ] 
+    if as_dict:
+      names = ['cxxxx','cyyyy','czzzz','cyzyz','cxzxz','cxyxy','cyyzz','cxxzz',
+               'cxxyy','cxxyz','cxxxz','cxxxy','cyyyz','cyyxz','cyyxy','czzyz',
+               'czzxz','czzxy','cyzxz','cyzxy','cxzxy']
+      return dict(zip(names, clist))
+    else:
+      return clist
+  else:
+    raise AssertionError('C has wrong shape')
+
+def extract_elements(C: Array) -> Dict:
+  """ Convert an elastic modulus tensor into a list of unique elements
+      
+      in 2d, these are:
+      cxxxx,cyyyy,cxyxy,cxxyy,cxxxy,cyyxy
+
+      in 3d, these are:
+      cxxxx,cyyyy,czzzz,cyzyz,cxzxz,cxyxy,cyyzz,cxxzz,cxxyy,cxxyz,cxxxz,cxxxy,
+      cyyyz,cyyxz,cyyxy,czzyz,czzxz,czzxy,cyzxz,cyzxy,cxzxy
+
+  Args:
+    C: A previously calculated elastic modulus tensor represented as an 
+      array of shape (spatial_dimension,spatial_dimension,spatial_dimension,
+      spatial_dimension), where spatial_dimension is either 2 or 3. C must 
+      satisfy both the major and minor symmetries, but this is not checked.
+  Return: a dict of the 6 (21) unique elastic constants in 2 (3) dimensions.
+  """
+  return _extract_elements(C,True)
+
+def extract_isotropic_moduli(C: Array) -> Dict:
+  """ There are a number of important constants used to describe the linear 
```

Comment:
Add a one sentence summary and remove one indentation from the rest of the comment maybe?

Reply:
Done.